src/bin/lttng-relayd/cmd-2-11.c

```diff
@@ -0,0 +1,104 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 51
+ * Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#define _LGPL_SOURCE
+#include <assert.h>
+
+#include <common/common.h>
+#include <common/sessiond-comm/relayd.h>
+
+#include <common/compat/endian.h>
+#include <common/compat/string.h>
+#include <lttng/constant.h>
+
+#include "cmd-2-11.h"
+#include "lttng-relayd.h"
+
+int cmd_create_session_2_11(const struct lttng_buffer_view *payload,
+		char *session_name, char *hostname,
+		uint32_t *live_timer, bool *snapshot)
+{
+	int ret;
+	struct lttcomm_relayd_create_session_2_11 header;
+	size_t header_len, received_names_size;
+	struct lttng_buffer_view session_name_view;
+	struct lttng_buffer_view hostname_view;
+
+	header_len = sizeof(header);
+
+	if (payload->size < header_len) {
+		ERR("Unexpected payload size in \"cmd_create_session_2_11\": expected >= %zu bytes, got %zu bytes",
+				header_len, payload->size);
+		ret = -1;
+		goto error;
+	}
+	memcpy(&header, payload->data, header_len);
+
+	header.session_name_len = be32toh(header.session_name_len);
+	header.hostname_len = be32toh(header.hostname_len);
+	header.live_timer = be32toh(header.live_timer);
+	header.snapshot = be32toh(header.snapshot);
+
+	received_names_size = header.session_name_len + header.hostname_len;
+	if (payload->size < header_len + received_names_size) {
+		ERR("Unexpected payload size in \"cmd_create_session_2_11\": expected >= %zu bytes, got %zu bytes",
+				header_len + received_names_size, payload->size);
+		ret = -1;
+		goto error;
+	}
+
+	/* Validate length against defined constant. */
+	if (header.session_name_len > LTTNG_NAME_MAX) {
+		ret = -ENAMETOOLONG;
+		ERR("Session name too long");
+		goto error;
+	}
+	if (header.hostname_len > LTTNG_HOST_NAME_MAX) {
+		ret = -ENAMETOOLONG;
+		ERR("Session name too long");
+		goto error;
+	}
+
+	/* Validate that names are (NULL terminated. */
+	session_name_view = lttng_buffer_view_from_view(payload, header_len,
+			    header.session_name_len);
+	hostname_view = lttng_buffer_view_from_view(payload,
+			header_len + header.session_name_len, header.hostname_len);
+
+	if (session_name_view.data[session_name_view.size - 1] != '\0') {
+		ERR("cmd_create_session_2_11 session_name is invalid (not NULL terminated)");
+		ret = -1;
+		goto error;
+	}
+
+	if (hostname_view.data[hostname_view.size - 1] != '\0') {
+		ERR("cmd_create_session_2_11 hostname is invalid (not NULL terminated)");
+		ret = -1;
+		goto error;
+	}
+
+	strcpy(session_name, session_name_view.data);
+	strcpy(hostname, hostname_view.data);
+
+	*live_timer = header.live_timer;
+	*snapshot = header.snapshot;
```

Comment:
`*snapshot = !!header.snapshot` since `header.snapshot` could contain any non-zero value.

Reply:
Not sure what you mean here. I get that !! will evaluate to either 0 or 1 but not sure about you comment regarding "could contain any non-zero value". Given that bool conversion is 0 -> 0, Anything else -> 1. Am I missing something?

Reply:
It's something I'm doing out of habit. I'd have to dig into the standard to see if an assignment from `uint32_t` to `bool` (implicit cast) is guaranteed to be safe.

I took the habit of using the double-negation since the standard says that:
```
An object declared as type _Bool is large enough to store the values 0 and 1.

http://port70.net/~nsz/c/c11/n1570.html#6.2.5p2
```

and that the result of a double-negation (`!!`) is guaranteed to be `0` or `1`
```
The result of the logical negation operator ! is 0 if the value of its operand compares unequal to 0, 1 if the value of its operand compares equal to 0. The result has type int. The expression !E is equivalent to (0==E).

http://port70.net/~nsz/c/c11/n1570.html#6.5.3.3p5
```

The implicit cast might be safe, I just couldn't find a definitive answer (not that I looked extensively).

Reply:
from you first link article 6.3.1.2 Boolean type (http://port70.net/~nsz/c/c11/n1570.html#6.3.1.2)

When any scalar value is converted to _Bool, the result is 0 if the value compares equal to 0; otherwise, the result is 1.